[22.05] Allow downloading individual metadata files in History

guerler opened the pull request:
Resolves #14123. This PR restores the dropdown download option displayed when individual files can be downloaded for a single history item.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

guerler committed:
Prep dropdown to account for meta files in dataset actions

guerler committed:
Add separate dataset download component

guerler committed:
Add metadata download options to dropdown

guerler committed:
Fix dropdown alignment and metadata download url

guerler committed:
Add test case for dataset download component

mvdbeek commented on client/src/components/History/Content/Dataset/DatasetDownload.vue:
```diff
@@ -0,0 +1,62 @@
+            return this.metaFiles && this.metaFiles.length > 0;
+        },
+        metaDownloadUrl() {
+            return prependPath(`dataset/get_metadata_file?hda_id=${this.item.id}&metadata_name=`);
```
This isn't the right path and fails with
```
uvicorn.error ERROR 2022-06-21 18:31:22,093 [pN:main,p:3277,tN:MainThread] Exception in ASGI application
Traceback (most recent call last):
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/uvicorn/protocols/http/h11_impl.py", line 366, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
    return await self.app(scope, receive, send)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/fastapi/applications.py", line 269, in __call__
    await super().__call__(scope, receive, send)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/starlette/applications.py", line 124, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/starlette_context/middleware/raw_middleware.py", line 96, in __call__
    await self.app(scope, receive, send_wrapper)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/starlette/middleware/base.py", line 69, in __call__
    await response(scope, receive, send)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/starlette/responses.py", line 260, in __call__
    await wrap(partial(self.listen_for_disconnect, receive))
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 662, in __aexit__
    raise exceptions[0]
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/starlette/responses.py", line 256, in wrap
    await func()
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/starlette/responses.py", line 245, in stream_response
    async for chunk in self.body_iterator:
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/starlette/middleware/base.py", line 58, in body_stream
    raise app_exc
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/starlette/middleware/base.py", line 36, in coro
    await self.app(scope, request.receive, send_stream.send)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/starlette/exceptions.py", line 93, in __call__
    raise exc
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/starlette/exceptions.py", line 82, in __call__
    await self.app(scope, receive, sender)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
    raise e
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/starlette/routing.py", line 670, in __call__
    await route.handle(scope, receive, send)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/starlette/routing.py", line 418, in handle
    await self.app(scope, receive, send)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/a2wsgi/wsgi.py", line 140, in __call__
    return await responder(scope, receive, send)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/a2wsgi/wsgi.py", line 179, in __call__
    raise self.exc_info[0].with_traceback(
  File "/Users/mvandenb/src/galaxy/lib/galaxy/web/framework/middleware/error.py", line 165, in __call__
    app_iter = self.application(environ, sr_checker)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/paste/recursive.py", line 85, in __call__
    return self.application(environ, start_response)
  File "/Users/mvandenb/src/galaxy/.venv/lib/python3.9/site-packages/paste/httpexceptions.py", line 640, in __call__
    return self.application(environ, start_response)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/web/framework/base.py", line 159, in __call__
    return self.handle_request(environ, start_response)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/web/framework/base.py", line 244, in handle_request
    body = method(trans, **kwargs)
TypeError: get_metadata_file() got an unexpected keyword argument 'metadata_name'
```

The right route should be `/api/histories/{history_id}/contents/{history_content_id}/metadata_file` with a parameter `metadata_file`.

guerler replied:
Interesting, did this change in comparison to the previous history?

mvdbeek replied:
No, but the old history is broken too.

mvdbeek replied:
I can take this over, we should restore the old route for legacy reasons and we should have a selenium test.

guerler replied:
Thank you @mvdbeek, let me know if you want me to make additions/adjustments.

mvdbeek committed:
Fix dataset controller ``get_metadata_file``

We need to keep the metadata_name argument, this broke in
https://github.com/galaxyproject/galaxy/pull/13348.

Also uses the API service code.

mvdbeek committed:
Use API route to download metadata files

mvdbeek committed:
Add selenium test for downloading metadata files
